pypeit/scripts/ql.py

```diff
@@ -0,0 +1,384 @@
+                    redux_path, 'Masters', 'Master*'))
+                if len(master_files) > 0 and not args.clobber_calibs:
+                    msgs.info('Master files already exist.  Skipping calibration.')
+                    continue
+                # Run
+                pypeIt = pypeit.PypeIt(calib_pypeit_file,
+                                       redux_path=redux_path, 
+                                       calib_only=True)
+                calib_dict = pypeIt.calib_all()
+        if args.calibs_only:
+            msgs.info("Calibrations only requested.  Exiting")
+            return
+
+        # Science files                                
+        if args.sci_files is not None:
+            sci_idx = np.in1d(ps.fitstbl['filename'], args.sci_files)
+        else:
+            sci_idx = ps.fitstbl['frametype'] == 'science'
+
+        if np.sum(sci_idx) == 0:
+            msgs.error('No science frames found in the provided files.  Add at least one or specify using --sci_files.')
+
+        # Generate science setup object
+        full_scifiles = [os.path.join(dir_path, sci_file) for dir_path, sci_file in zip(
+            ps.fitstbl['directory'][sci_idx], ps.fitstbl['filename'][sci_idx])]
+        ps_sci = pypeitsetup.PypeItSetup.from_rawfiles(
+                full_scifiles, ps.spectrograph.name)
+        ps_sci.run(setup_only=True, no_write_sorted=True)
+        # Limit to a single science setup
+        if len(ps_sci.fitstbl.configs.keys()) > 1:
+            msgs.error('Your science files come from more than one setup. Please reduce them separately.')
+
+        # Masters dir and their setup
+        if args.masters_dir is None:
+            # Match to calibs
+            _, masters_dir = match_science_to_calibs(
+                ps_sci, calib_dir)
+        else:
+            masters_dir = args.masters_dir
+
+        # Parse for setup
+        master_files = glob.glob(os.path.join(
+            masters_dir, 'Master*'))
+        if len(master_files) == 0:
+            msgs.error('No Master files found in {:s}'.format(masters_dir))
+        masters_setup_and_bit = os.path.basename(
+            master_files[0]).split('_')[1:3]
+
+        # Build the PypeIt file and link to Masters
+        sci_pypeit_file, _ = \
+                generate_sci_pypeitfile(
+                    args.redux_path, 
+                    full_scifiles, 
+                    masters_dir, 
+                    masters_setup_and_bit, 
+                    ps_sci, 
+                    maskID=args.maskID, 
+                    slitspatnum=args.slitspatnum,
+                    det=args.det,
+                    boxcar_radius=args.boxcar_radius,
+                    stack=args.stack)
+        
+        # Run it
+        redux_path = os.path.dirname(sci_pypeit_file)  # Path to PypeIt file
+        run_pargs = run_pypeit.RunPypeIt.parse_args(
+            [sci_pypeit_file, '-r', redux_path])
+        run_pypeit.RunPypeIt.main(run_pargs)
```

Comment:
Use the `PypeIt` class directly instead of the script.

Reply:
agreed. done